### PR TITLE
Release package compatibility

### PR DESCRIPTION
**Pull request by tuffnatty**
This is a long patch set, and probably it could be better structured. What it gives is:

- Ubuntu 16.04+ compatible build with Qt 5.15.9 and ICU 72.1
- Windows builds with Qt 5.15.9
- MacOS 10.13+ compatible build with Qt 5.15.8
- MacOS 10.12 compatible build with Qt 5.13.2
- optional package with embedded Python on MacOS

### CODE REVIEW

**Comment by pawelsalawa** [comment]
Impressive piece of work. As it's WIP, I'll hold with mering.

**Comment by tuffnatty** [comment]
@pawelsalawa Generally, I am satisfied with the resulting builds. You can see and test them at https://github.com/tuffnatty/sqlitestudio/actions .

The Linux build has multiplied: now there are 5 workflows you have to run in sequence if running for the first time. The new four workflows build the runner container and SQLiteStudio's bundled dependencies. Probably the dependency workflows should be merged in one (Qt takes an hour, the others take a couple of minutes each) but it has been easier to debug separately so far.
The final Linux release build produces loads of packages using build matrix, but actually only glibc_2.23/Qt 5.15.9 is the one I am suggesting to keep. glibc_2.31/Qt 5.15.2 is basically what we were having lately with the Ubuntu 20.04 runner.

Bundling Python 3.9 on MacOS is a questionable decision. If we bundle it ourselves why not 3.11? Did not try it yet but 3.10 worked well on one occasion. For the slim package, I think it would be better to let user choose their desired Python shared library in configuration and link to it dynamically (discussed in #4662), but if it has not become possible with the recent versions, I would suggest shipping 3 plugin versions.

Keeping the numerous Qt's dependency versions in the MacOS workflow may be boring, but until the distfiles cache is cleared it should not be necessary. Probably can also be automated.

**Comment by pawelsalawa** [comment]
Will the Qt have to be built more than once? I imagine it would be required only if anything changes in Qt, right? (like version, or patches applied)

Python 3.11 - the problem here will be C API or ABI compatibility. I don't remember exactly right now, because it's been a while now since I worked on Python plugin, but I remember that I was surprised, that the API compatibility is not out of box when going with newer versions. It's something worth trying and verifying (if the plugin loads to SQLiteStudio, it should be fine).

**Comment by tuffnatty** [comment]
Yes, the plan is to keep the build cached until Qt version change or new patches. I also think the old build artifacts are occasionally cleaned automatically, and an infrequent scheduled rebuild should help here.

**Comment by tuffnatty** [comment]
@pawelsalawa I have included the small dependencies (ICU, OpenSSL, and Tcl) to the container image, leaving us with just 3 Linux workflows and 10 commits to review.

**Comment by pawelsalawa** [comment]
While I admire the effort and beauty of having various variants of builds (different Qt versions, different glibc versions on Linux), is it really profitable? I mean since glibc_2.23 will work on both older and newer Linux (if I understand it correctly), then is it worth to build the others, which are not as portable? Also if we can have Qt 5.15.9, why keeping 5.15.2?

If there are good reasons, I'm okay with it. Otherwise I think it will be waste of space and GitHub actually does limit it (it charges a fee for everything over 500MB currently stored in all action artifacts). I have a cleanup job in place, but having so many artifacts has significant impact on how long artifacts can be kept.

We can keep all the action code that builds other variations as commented code - just in case.

What do you think?

File: .github/workflows/lin_qt.yml [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+        strategy:
+          fail-fast: false
+          matrix:
+            binary_compatibility:
+              - "glibc_2.23"  # Ubuntu 16.04
+              - "glibc_2.27"  # Ubuntu 18.04
+              - "glibc_2.31"  # Ubuntu 20.04, Debian 11
+            version:
+              - "5.15.9"
```
Comment (pawelsalawa) [inline .github/workflows/lin_qt.yml:35]: Is it possible to pull this Qt version string to header of the action, so it's more like a declared variable?

File: .github/workflows/lin_qt.yml [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+                    curl -L https://download.qt.io/official_releases/qt/5.15/${{ matrix.version }}/submodules/$_m-everywhere-opensource-src-${{ matrix.version }}.tar.xz \
+                    | xzcat \
+                    | tar -x
+                done
+
+            - name: Apply patches
+              run: |
+                set -x
+                cd qtbase-everywhere-src-${{ matrix.version }}
+                case ${{ matrix.version }} in 5.15.9)
```
Comment (pawelsalawa) [inline .github/workflows/lin_qt.yml:107]: If Qt version is declared in top of the action, we can re-use it here, right?

File: .github/workflows/lin_qt.yml (hunk 1 above)
Comment (pawelsalawa) [reply .github/workflows/lin_qt.yml:35]: Actually, nevermind. I see the idea behind it now. I think we can keep it here, as is.

File: .github/workflows/lin_qt.yml (hunk 2 above)
Comment (pawelsalawa) [reply .github/workflows/lin_qt.yml:107]: Forget this one too.

**Comment by pawelsalawa** [comment]
As I go through the lin_release now, I think that removing the Qt 5.15.2 completely from the workflow would significantly simplify this workflow and - if I'm not mistaken - would enable the Qt 5.15.9 to be included in docker images, thus making it more persistent across subsequent release builds, right?

**Comment by tuffnatty** [comment]
@pawelsalawa Of course it's not directly profitable but helps to ensure the builds are compatible with a variety of environments. I have just removed some extra Linux builds leaving only Xenial/Qt-5.15.9 and Bionic/Qt-5.15.2. The 5.15.2 build uses the same Qt binaries provided by aqtinstall as the old build, so I thought it can probably help to catch any problems with the new build during the transition period. Naturally, removing it will make the code a bit cleaner.
As for the GitHub storage limits, are you sure it applies to public repositories?

File: .github/workflows/mac_release.yml [hunk 3]
```diff
@@ -27,32 +27,72 @@ jobs:
     build:
         runs-on: macos-11
 
+        strategy:
+          fail-fast: false
+          matrix:
+            binary_compatibility:
+              - "darwin_16.x86_64"  # macOS 10.12, the earliest version supported by QT 5.13
```
Comment (pawelsalawa) [inline .github/workflows/mac_release.yml:34]: Do you think it's worth supporting 9-years old system? I'm asking this, cause this implies going back to Qt 5.13, whereas in SQLiteStudio 3.5 I plan to upgrade to Qt 6.5 and not looking back from then. I mean if someone will need to use SQLiteStudio on macOS 10.13 there will be SQLiteStudio 3.4.x, assuming the principle: the old OS, the old software. I mean I'm happy to support older systems, as long as it doesn't hold back the progress of the project. Sticking to supporting older Qt versions is holding back the progress.

Comment (tuffnatty) [reply .github/workflows/mac_release.yml:34]: I would not be touching it if that problem did not hit me before (around version 3.3.0). I think there are many Mac, Linux, and Windows users that can't upgrade or won't upgrade for a lot of different reasons. I am mostly on MacOS 10.14 now, for example, but keep a 10.12 machine around. As for the support, no changes to the code are needed at the moment (ok, hung QtScript interpereters can't be terminated, but it's their own choice). Also, Qt 5.12 is specified as the minimum supported version somewhere in docs. Also, Windows 7 is somewhat 14 years old and is still supported by MinGW-W64, Qt 5 and SQLiteStudio.

**Comment by pawelsalawa** [comment]
@tuffnatty Okay, that makes sense. Let's keep it for now. 

And you're right about no limits for public repos. I wasn't aware of that.

File: .github/workflows/mac_release.yml (hunk 3 above)
Comment (tuffnatty) [reply .github/workflows/mac_release.yml:34]: And... I think MacOS 10.12 users constitute around 8% of total MacOS users ;-) (source: https://ports.macports.org/port/curl/stats/ ; sadly there is no better data as Apple misreports at least versions 10.15+ on web stats)

Comment (tuffnatty) [reply .github/workflows/mac_release.yml:34]: If the plan is to jump straight to 6.5, lots of old systems will be dropped. Qt 6.4 is at least 10.14+ compatible; Qt 6.5 is MacOS 11+ which makes it a no-go for me.

**Comment by pawelsalawa** [comment]
There's an issue with builds for Windows with Qt 5.15.9. At least builds I took from your fork. For 32 bit version I get following error when I try to run it:
image
and for 64 bit I got 2 errors one after another:
image
image

Windows builds with Qt 5.15.2 work fine.

If Windows is not your daily OS, I can do the smoke tests (like this one) for you. Not a problem.

**Comment by tuffnatty** [comment]
Well, the Windows builds definitely used to start at one moment, but then something changed... will check soon.

File: .github/workflows/mac_release.yml (hunk 3 above)
Comment (pawelsalawa) [reply .github/workflows/mac_release.yml:34]: If I may ask - why are you not upgrading to newer versions? I have a quite old Mac by myself (which I use very, very rarely) and like a year ago or so I upgraded to Monterey. For free and without issues.

I'm curious why to stick to older OS versions?

**Comment by tuffnatty** [comment]
Windows builds are working again.

File: .github/workflows/mac_release.yml (hunk 3 above)
Comment (tuffnatty) [reply .github/workflows/mac_release.yml:34]: I even googled what wise men answer to this question. The leading experts say: "Intellectual laziness, mostly".

Another way to call it is lack of time for foolishness: I know my tools well, though dated they are working well; for the new versions, I have to relearn, they are bloated, the new functionality they give (if any) is not useful for me, and often they break backward compatibility. The major OS version upgrades almost never go very smooth. OS X 10.9, Windows XP (in a VM) and Ubuntu 16.04 were good enough for me, the new versions never ran faster or increased backward compatibility; on the opposite, they break things (Linux systemd story), unconditionally promote my filesystem to another (slower and very young) filesystem type (MacOS 10.14+), or push unconditional updates (Windows 10+) and add a lot of telemetry (Windows 10+ and MacOS 11+).

Comment (tuffnatty) [reply .github/workflows/mac_release.yml:34]: 6.2 has no compatibility advantages compared to 6.4, so 6.4 seems good enough for me.

**Comment by pawelsalawa** [comment]
I confirm that now the builds work.

I see you removed the WIP prefix. Shall I do one more review and merge, or do you foresee some additional commits?

**Comment by tuffnatty** [comment]
1) I see a new ICU release but I think it's better to keep a single version on all platforms for now.
2) You suggested including Qt build to the container build. It's very easy but then Qt will have to be rebuilt when upgrading any of the other dependencies. Not a great problem, of course. I can do it if you wish.

**Comment by pawelsalawa** [comment]
Since we agreed to support more than one Qt version in release packages, I think we should stay with Qt outside of the container. In that case I will have one last look (not now, but soon) and merge.

This is really great piece of work you've done here. Thank you!

**Comment by tuffnatty** [comment]
> @tuffnatty Hi, would you mind checking why some Windows builds failed? https://github.com/pawelsalawa/sqlitestudio/actions/runs/4910709807 and https://github.com/pawelsalawa/sqlitestudio/actions/runs/4910827559
> 
> You may probably have better/quicker guess. I took a look, but I have no quick answer.

See PR #4792. I also do not quite understand what was happening there but It got fixed by switching to msys2 bash for compiling.

**Pull request merged**